### PR TITLE
Add command to create new personal access token

### LINKED ISSUE #30: Add command to create personal access tokens
Add a console command to create personal access tokens

### PR DESCRIPTION
**Pull request by nunorafaelrocha**
Closes #30.

### CODE REVIEW

File: lib/Bitreserve/BitreserveClient.php [hunk 1]
```diff
@@ -215,6 +215,29 @@ public function getUser()
     }
 
     /**
+     * Creates a new Personal Access Token (PAT).
```
Comment (fixe) [inline lib/Bitreserve/BitreserveClient.php]: Use singular please: `create`.

File: lib/Bitreserve/BitreserveClient.php [hunk 2]
```diff
@@ -215,6 +215,29 @@ public function getUser()
...
+     *
+     * @param  string $login       Login email or username.
+     * @param  string $password    User password.
+     * @param  string $description PAT description or label.
```
Comment (fixe) [inline lib/Bitreserve/BitreserveClient.php]: Remove `or label`.

File: lib/Bitreserve/BitreserveClient.php [hunk 3]
```diff
@@ -215,6 +215,29 @@ public function getUser()
     }
 
     /**
+     * Creates a new Personal Access Token (PAT).
+     *
+     * @param  string $login       Login email or username.
+     * @param  string $password    User password.
```
Comment (fixe) [inline lib/Bitreserve/BitreserveClient.php]: `Password`.

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        // Form variables.
+        $this->description = $this->getDescription();
+        $this->otp = null;
+        $this->password = null;
+        $this->username = null;
+
+        // Get user credentials.
+        $this->getCredentialsFields();
+
+        $newPAT = null;
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: `$pat`.

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+use Symfony\Component\Process\Process;
+
+class CreateTokenCommand extends Command
+{
+    /**
+     * Configure the command.
+     */
+    protected function configure()
+    {
+        $this->setName('token:create')
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: I suggest renaming this to `tokens:create`. Keep in mind that we will add `tokens:list` and `tokens:revoke`.

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        // Command line output interface.
+        $this->output = $output;
+
+        // Bitreserve Client.
+        $this->client = new BitreserveClient();
+
+        // Form variables.
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: `Form`? :crying_cat_face:

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        $this->description = $this->getDescription();
+        $this->otp = null;
+        $this->password = null;
+        $this->username = null;
+
+        // Get user credentials.
+        $this->getCredentialsFields();
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: `getCredentials`.

File: lib/Bitreserve/console.php [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+require 'vendor/autoload.php';
+
+require 'Command/CreateTokenCommand.php';
+
+use Symfony\Component\Console\Shell;
```
Comment (fixe) [inline lib/Bitreserve/console.php]: Sort.

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+            }
+
+            return $value;
+        });
+    }
+
+    /**
+     * Gets user account verification code (OTP).
+     *
+     * @return void
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: Is this correct?

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+     */
+    public function getOtp()
+    {
+        $dialog = $this->getHelperSet()->get('dialog');
+
+        return $dialog->askAndValidate($this->output, '<question>Authorization code:</question> ', function ($value) {
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: `<question>Verification code:</question>` to match bitreserve.org copy.

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+            if (null === $value || '' === $value) {
+                return $this->getPassword();
+            }
+
+            return $value;
+        });
+    }
+
+    /**
+     * Gets user account verification code (OTP).
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: `Get`.

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+            if (null === $value || '' === $value) {
+                return $this->getLogin();
+            }
+
+            return $value;
+        });
+    }
+
+    /**
+     * Gets user password.
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: `Get`.

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+            if (null === $value || '' === $value) {
+                return $this->getDescription();
+            }
+
+            return $value;
+        });
+    }
+
+    /**
+     * Gets user login (username or email).
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: `Get`.

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+     *
+     * @return void
+     */
+    public function getLogin()
+    {
+        $dialog = $this->getHelperSet()->get('dialog');
+
+        return $dialog->askAndValidate($this->output, '<question>Username/Email:</question> ', function ($value) {
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: Rename to `<question>Username or Email:</question>`.

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+     *
+     * @return array The new personal access token.
+     */
+    public function requestToken()
+    {
+        return $this->client->createToken($this->username, $this->password, $this->description, $this->otp);
+    }
+
+    /**
+     * Gets login and password.
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: `Get`.

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+     * @return void
+     */
+    public function getCredentialsFields()
+    {
+        $this->username = $this->getLogin();
+        $this->password = $this->getPassword();
+    }
+
+    /**
+     * Gets PAT description from the output.
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: `Get`.

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+        // Process new PAT.
+        while (null === $newPAT) {
+            try {
+                $newPAT = $this->requestToken();
+            } catch (AuthenticationRequiredException $e) {
+                $output->writeln('Bad credentials, please try again.');
+
+                $this->getCredentialsFields();
+            } catch (TwoFactorAuthenticationRequiredException $e) {
+                $output->writeln('Two factor authentication is enabled on this account, please enter you verification code.');
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: `the verification code`.

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+        $this->getCredentialsFields();
+
+        $newPAT = null;
+
+        // Process new PAT.
+        while (null === $newPAT) {
+            try {
+                $newPAT = $this->requestToken();
+            } catch (AuthenticationRequiredException $e) {
+                $output->writeln('Bad credentials, please try again.');
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: `Wrong credentials` perhaps?

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+    /**
+     * Configure the command.
+     */
+    protected function configure()
+    {
+        $this->setName('tokens:create')
+            ->setDescription('Create a new Personal Access Token');
+    }
+
+    protected function initialize(InputInterface $input, OutputInterface $output)
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: Missing comment.

File: lib/Bitreserve/Command/CreateTokenCommand.php [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+        // Command line output interface.
+        $this->output = $output;
+
+        // Bitreserve Client.
+        $this->client = new BitreserveClient();
+
+        // Input variables.
+        $this->description = null;
+        $this->otp = null;
+        $this->password = null;
```
Comment (fixe) [inline lib/Bitreserve/Command/CreateTokenCommand.php]: Sort.

**Comment by fixe** [comment]
Please add an example to the `README` file.

File: README.md [hunk 21]
```diff
@@ -49,10 +49,16 @@ Require the library package as a dependency:
...
 ## Basic usage
 
-In order to create a Personal Access Token (PAT) and learn more about the Bitreserve API, please check out the [developer website](https://developer.bitreserve.org/api/v0/#authentication).
+In order to learn more about the Bitreserve API, please check out the [developer website](https://developer.bitreserve.org/api/v0/#authentication).
+
+First, create a Personal Access Token (PAT) using the command described below.
 
 Then, create a new instance of the `Client` class with token. Take a look at the following examples and explore more on the [examples](https://github.com/bitreserve/bitreserve-sdk-php/tree/master/examples) directory.
 
+### Create a Personal Access Token
```
Comment (fixe) [inline README.md]: Move this to L54.

**Comment by fixe** [comment]
:facepunch:

**Comment by nunorafaelrocha** [comment]
:bowtie:

**Pull request merged**